File: etrike_protocol/codecs/typescript/pwt.ts
import { type CanFrame, type DecodeResult, type EncodeResult, frame, validateFrame } from "./types";

export const PWT_BUS = "powertrain";
export const PWT_DCDC_COMMAND_ID = 0x10262b27;
export const PWT_DLC = 8;

export function encodePwtDcdcCommand(values: Readonly<Record<string, unknown>>): EncodeResult {
  const control = values.control ?? 1;
  const reset = values.reset_control ?? 0;
  if (typeof control !== "number" || typeof reset !== "number" || ![0, 1].includes(control) || ![0, 1].includes(reset)) {
    return ["invalid_enum"];
  }
  return ["ok", frame(PWT_BUS, PWT_DCDC_COMMAND_ID, "extended", [control, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, reset])];
}

export function decodePwtDcdcCommand(input: CanFrame): DecodeResult<Record<string, number>> {
  const status = validateFrame(input, {
    bus: PWT_BUS,
    id: PWT_DCDC_COMMAND_ID,
    frameFormat: "extended",
    dlc: PWT_DLC,
  });
  if (status !== "ok") return [status];
  if (![0, 1].includes(input.data[0]) || ![0, 1].includes(input.data[7])) return ["invalid_enum"];
  if (input.data.subarray(1, 7).some((value) => value !== 0xff)) return ["constant_mismatch"];
  return ["ok", {
    control: input.data[0],
    reserved_1: 0xff,
    reserved_2: 0xff,
    reserved_3: 0xff,
    reserved_4: 0xff,
    reserved_5: 0xff,
    reserved_6: 0xff,
    reset_control: input.data[7],
  }];
}

File: autoware/src/our_packages/etrike_protocol/codecs/python/pwt.py
from __future__ import annotations

from collections.abc import Mapping

from .types import CodecStatus, Frame, validate_frame

BUS = "powertrain"
DCDC_COMMAND_ID = 0x10262B27
DLC = 8


def encode_dcdc_command(values: Mapping[str, object]) -> tuple[CodecStatus, Frame | None]:
    control = values.get("control", 1)
    reset = values.get("reset_control", 0)
    if not isinstance(control, int) or isinstance(control, bool):
        return "invalid_enum", None
    if not isinstance(reset, int) or isinstance(reset, bool):
        return "invalid_enum", None
    if control not in (0, 1) or reset not in (0, 1):
        return "invalid_enum", None
    return "ok", Frame(BUS, DCDC_COMMAND_ID, "extended", bytes([control, *([0xFF] * 6), reset]))


def decode_dcdc_command(frame: Frame) -> tuple[CodecStatus, dict[str, int] | None]:
    status = validate_frame(
        frame, bus=BUS, can_id=DCDC_COMMAND_ID, frame_format="extended", dlc=DLC
    )
    if status != "ok":
        return status, None
    if frame.data[0] not in (0, 1) or frame.data[7] not in (0, 1):
        return "invalid_enum", None
    if any(value != 0xFF for value in frame.data[1:7]):
        return "constant_mismatch", None
    return "ok", {
        "control": frame.data[0],
        **{f"reserved_{index}": 0xFF for index in range(1, 7)},
        "reset_control": frame.data[7],
    }
